File: python/cube.py
from dataclasses import dataclass
import sys


@dataclass
class Game:
    id: int
    red: int
    green: int
    blue: int


def capability(g: Game) -> int:
    return g.id if g.red <= 12 and g.green <= 13 and g.blue <= 14 else 0

def power(g: Game) -> int:
    return g.red * g.green * g.blue

def parse_round(s: str) -> tuple[int, int, int]:
    red, green, blue = 0, 0, 0

    for chunk in s.split(","):
        [v, k] = chunk.strip().split(" ")
        match k:
            case "blue":
                blue = int(v)
            case "green":
                green = int(v)
            case "red":
                red = int(v)
            case _:
                raise Exception

    return (red, green, blue)


def parse(line: str) -> Game:
    [g, rest] = line.split(":")
    [_, id] = g.split(" ")
    red, green, blue = 0, 0, 0

    for s in rest.split(";"):
        r, g, b = parse_round(s)
        red = max(red, r)
        green = max(green, g)
        blue = max(blue, b)

    return Game(int(id), red, green, blue)


def main() -> None:
    filename = sys.argv[1]
    with open(filename) as f:
        lines = [x.strip() for x in f.readlines()]
        results = sum([capability(parse(x)) for x in lines])
        powers = sum([power(parse(x)) for x in lines])
        print(f"{results=}, {powers=}")


if __name__ == "__main__":
    main()

File: go/cube.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

type game struct {
	id    int
	red   int
	green int
	blue  int
}

func (g *game) capability() int {
	if g.red <= 12 && g.green <= 13 && g.blue <= 14 {
		return g.id
	}
	return 0
}

func (g *game) power() int {
	return g.red * g.green * g.blue
}

func parse_round(s string) (int, int, int) {
	var red, blue, green, v int
	var err error

	for _, chunk := range strings.Split(s, ",") {
		components := strings.Split(strings.TrimSpace(chunk), " ")
		if v, err = strconv.Atoi(components[0]); err != nil {
			panic("can't parse")
		}

		switch components[1] {
		case "blue":
			blue = v
		case "green":
			green = v
		case "red":
			red = v
		default:
			panic("invalid colour")
		}
	}

	return red, green, blue
}

func extract_id(chunk string) int {
	var id int
	var err error

	if id, err = strconv.Atoi(strings.Split(chunk, " ")[1]); err != nil {
		panic("can't parse")
	}

	return id
}

func parse(line string) game {
	csplit := strings.Split(line, ":")
	id := extract_id(csplit[0])
	var red, green, blue int

	for _, s := range strings.Split(csplit[1], ";") {
		r, g, b := parse_round(s)
		red = max(red, r)
		green = max(green, g)
		blue = max(blue, b)
	}

	return game{id: id, red: red, green: green, blue: blue}
}

func main() {
	var result int
	var power int

	filename := os.Args[1]
	file, err := os.Open(filename)

	if err != nil {
		panic("can't find file")
	}

	defer file.Close()

	scanner := bufio.NewScanner(file)

	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())
		game := parse(line)
		result += game.capability()
		power += game.power()
	}

	fmt.Printf("results=%d, powers=%d\n", result, power)
}
